### PR TITLE
plugins.tv999: new plugin for http://tv999.bg/live.html

### LINKED ISSUE #2905: [New Plugin] Add TV999
## Plugin Request
- [x] This is a plugin request and I have read the contribution guidelines and plugin request requirements.

### Description
Bulgarian TV channel that mainly features Russian content with Bulgarian subtitles.

### Example stream URLs
http://tv999.bg/live.html

### PR DESCRIPTION
**Pull request by mkbloke**
closes #2905

### CODE REVIEW

**Comment by gravyboat** [comment]
@mkbloke Let's be careful about adding too many plugins. When people say content is "popular" I want to see actual stats showing the site is popular. As well as ensuring it's not just stolen content or VODs crushed together in to a live stream, since that doesn't really count.

**Comment by mkbloke** [comment]
@gravyboat, fair points, for sure.  I have done some digging following your comment and have found that it *seems* to be a live channel available via satellite and cable in Bulgaria, according to these pages (via translation in Chrome):

- http://tv999.bg/%D0%B7%D0%B0-%D0%BD%D0%B0%D1%81.html
- https://bg.wikipedia.org/wiki/999_%D0%A2%D0%92

As for popularity, that seems harder to ascertain.  :-/

**Comment by gravyboat** [comment]
@mkbloke Yeah when sites don't show viewers like many live streaming websites do it makes it hard to determine. The real problem I've been finding with people submitting these sites lately is what we consider to be "live" and people are trying to skirt the rules to get their crappy VOD site in because "but this is live TV in country/state XYZ!" which isn't live at all and I don't want to add more of these styles of plugins.

**Comment by roshavagarga** [comment]
@mkbloke @gravyboat The plugin works and I can confirm that it's a small TV station that is available nationally - I believe it was from the city of Yambol, might be a former regional TV that went from being public to being private. I'd say the amount of people using it wouldn't be high, but the amount of work it'll require will be minimal, as the website hasn't been changed in ages.

It's an entirely entertainment channel - a few documentaries here and there, but mostly movies and TV series, often foreign ones with Bulgarian subtitling.

File: src/streamlink/plugins/tv999.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+
+class TV999(Plugin):
+    url_re = re.compile(r'https?://(?:www\.)?tv999\.bg/live\.html')
+    iframe_re = re.compile(r'<iframe.*src="([^"]+)"')
+    hls_re = re.compile(r'src="([^"]+)"\s+type="application/x-mpegURL"')
+
+    iframe_schema = validate.Schema(
+        validate.transform(iframe_re.search),
+        validate.get(1),
+        validate.any(None, validate.url()),
```
Comment (bastimeyer) [inline src/streamlink/plugins/tv999.py:16]: If the regex search fails, it'll return a `None` value, and trying to access `[1]` on a `None` value will raise a `TypeError`. That's what the `validate.any(None, validate.all(...))` call is for.

File: src/streamlink/plugins/tv999.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    )
+
+    hls_schema = validate.Schema(
+        validate.transform(hls_re.search),
+        validate.get(1),
+        validate.any(None, validate.all(
```
Comment (bastimeyer) [inline src/streamlink/plugins/tv999.py:22]: Same here

**Comment by mkbloke** [comment]
@roshavagarga, I've just pushed some small cosmetic changes to the plugin, so if you could test again, it would be much appreciated.

By the way, I deleted the comment I made in your plugin request issue because I wanted to make another comment with the link to the newly pushed commit, rather than as a result of the conversation here.

Unfortunately, I can no longer seem to test directly myself.  I was using Tor configured with an exit node in Bulgaria, but perhaps I just got lucky yesterday as this no longer seems to be working for me now.  If the current commit works for you, I'll leave it as is and hopefully it can be merged, given the info you have provided above.

Cheers.

File: src/streamlink/plugins/tv999.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+            validate.transform(lambda x: "http:{0}".format(x)),
```
Comment (bastimeyer) [inline src/streamlink/plugins/tv999.py:23]: use `update_scheme`:

```python
from streamlink.utils.url import update_scheme
update_scheme("http:", "//foo.bar/baz") == "http://foo.bar/baz"
update_scheme("http:", "http://foo.bar/baz") == "http://foo.bar/baz"
update_scheme("http:", "https://foo.bar/baz") == "https://foo.bar/baz"
```

File: src/streamlink/plugins/tv999.py [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+from streamlink.plugin import Plugin
+from streamlink.plugin.api import validate
+from streamlink.stream import HLSStream
+
+
+class TV999(Plugin):
+    url_re = re.compile(r'https?://(?:www\.)?tv999\.bg/live\.html')
+    iframe_re = re.compile(r'<iframe.*src="([^"]+)"')
+    hls_re = re.compile(r'src="([^"]+)"\s+type="application/x-mpegURL"')
```
Comment (bastimeyer) [inline src/streamlink/plugins/tv999.py:11]: Just saying this in general, and it isn't strictly meant in regards to your changes, but maybe we should stop using those wonky regexes for extracting data from HTML nodes and instead use a proper XML/HTML parser to get the nodes via xpath queries or CSS selectors and then read the attribute values. But ignore this now, just wanted to mention it.

**Review by bastimeyer — CHANGES_REQUESTED** [review]

**Comment by mkbloke** [comment]
@gravyboat, I see that roshavagarga has another plugin request open (#2906) which looks fairly reasonable to me, but the above conversation makes me unsure about creating a plugin for it.

I also wonder about sites like Mico, which I created a plugin for in #3188.  There's another plugin request (much older) for what appears to be a similar chat site in issue #2331.  Obviously these are live streaming sites, but it could be a case that these will pop up more and more often, which I guess might result in reluctance to accept streamlink plugins for them.

Perhaps some of these old plugin request issues should be reviewed again and perhaps closed.  It might also help if additional labels were used on these plugin request issues, say ```pending review``` and ```approved``` to indicate which ones are actually going to be accepted and merged into master if they are created?

**Comment by bastimeyer** [comment]
> I see that roshavagarga has another plugin request open

Nothing unusual... :thinking:
https://github.com/streamlink/streamlink/issues?q=is%3Aissue+author%3Aroshavagarga+sort%3Aupdated-desc+label%3A%22plugin+request%22

There have been a ton of similar requests made by him over the past years and most of those small Bulgarian TV streams are covered by the `cdnbg` plugin. It's always the same or a similar structure:
https://github.com/streamlink/streamlink/blob/4fa902fc93470af30616e29e0fd49692a05ae84c/src/streamlink/plugins/cdnbg.py

Looks like this new plugin might be a duplicate and it could be added to the `cdnbg` plugin.

**Comment by mkbloke** [comment]
Maybe I should just close this PR and consider trying to add this site to the cdnbg plugin, which as it is now seems to require a fix and update as per #3201...

**Comment by mkbloke** [comment]
@bastimeyer, I've looked at the cdnbg plugin and that seems designed specifically for finding IFRAMEs that are sourced from cdn.bg.  While I could modify the IFRAME regex to include sources from elsewhere, would that really be the right thing to do?

**Pull request merged**